README.md
# Islands skin for stylobate [![Build Status](https://travis-ci.org/yandex-ui/stylobate-islands.png?branch=master)](https://travis-ci.org/yandex-ui/stylobate-islands)

CHANGELOG.md
---
# Stylobate-islands Changelog
---

## v1.0.0 (in development)

- Added islands-v3 switch.

- Added islands-v3 input, with a lot of upgrades from the v2 one (except for the overall wrote-from-scratching):

    - `large` and `extra-large` sizes, with aliases for all of them.
    - Complex input made into a skin, triggered by having `__controller` element defined.

- Added islands-v3 button, with a lot of upgrades from the v2 one (except for the overall wrote-from-scratching):

    - Added `extra-large` button size.
    - Added `dark` and `pseudo-dark` button types.
    - Added shortcut aliases for button sizes (`s`, `xl` etc.).
    - Added button using two elements (when using `content` element).
    - Added `shape` modifier to button.

- Changed line-height on body to unitless.
- Changed `rem` to `px` for standalone variant.
- Added `$islands_px` for modifying variables using the same unit that is used for other variables, if necessary.
- Changed notation for elements from `foo-element` to `foo__element`.
- Removed all the deprecated code.
- Removed the “standalone” version.
- Using `islands-` prefix everywhere (docs and tests).
- Removed `tail` param from arrow in favor of `__helper`.
- Using Stylobate 1.0.0

## v0.28.0 (2014-07-03)

- Updated Stylobate and Stylus deps.

## v0.27.5 (2014-06-18)

- Updated Stylobate dep, fixes rem problem.

## v0.27.4 (2014-05-30)

- Updated Stylobate dep, fixes icon problems.

## v0.27.3 (2014-05-30)

- Fixed clean-css issue in promo-button, #74.
- Updated dev deps and fixed Travis issue.

## v0.27.2 (2014-04-26)

- Fixed broken `init_islands_root`.
- Updated Stylobate to v0.24.1.

## v0.27.1 (2014-04-24)

- Fixed a few left issues: links and form.

## v0.27.0 (2014-04-24)

- Applied latest Stylobate, some API changes, caching and other stuff.

## v0.26.1 (2014-04-17)

- Fixed rgba to rgba-ie in border-color values, #70.

## v0.26.0 (2014-04-10)

- Added a way to configure border-alpha of pseudo-buttons.
- Added a way to center the content of simple inputs.
- Fixed the difference in box-shadow for promo-buttons.
- Changed default text color to black, #67.
- Updated Stylus to `0.43` and tests for it.

## v0.25.1 (2014-03-31)

- Updated `if-ie.styl` to `0.5.0`.

## v0.25.0 (2014-03-26)

- Added vertical groups.
- Refactored groups to have placeholders.

## v0.24.1 (2014-03-06)

- Made inverted pseudo-button slightly brighter.
- Fixed bounding box of check icon.
- Added way to make a small placeholder.
- Refactored and tested variables.

## v0.24.0 (2014-03-06)

- Added `_inverted` modifier to pseudo-button.

- Added ie fallback for arrow skin

## v0.23.3 (2014-02-26)

- Updated Stylobate to `0.22.5`.

## v0.23.2 (2014-02-26)

- Fixed incorrect call of button skin in checkbox.

## v0.23.1 (2014-02-26)

- Updated Stylobate to `0.22.4`.

## v0.23.0 (2014-02-22)

- Added offset arguments for arrow icon.
- Removed disabled guards, as it is now have no events.
- Using stylobate's disabled kind for disabled skin.
- Fixed selectors for file inputs a bit.
- Updated deps.

## v0.22.0 (2014-02-11)

- Added arrow icon skin.

## v0.21.1 (2014-01-23)

- Updated Stylobate.

## v0.21.0 (2014-01-20)

- Updated Stylobate.
- Translated links to the new Stylobate.

## v0.20.0 (2014-01-17)

- Added placeholder for fields.
- Updated tests with proper content (after new CSSComb and Autoprefixer).

## v0.19.1–v0.19.2 (2014-01-10)

- Some minor hotfixes.

## v0.19.0 (2014-01-10)

- Now using require instead of import.
- Refactored some skins using the latest Stylobate features.
- Updated deps.

## v0.18.2 (2013-12-12)

- Some enhancements for menu-item (ellipsis param, `wsnw`, no user select, separate size).

## v0.18.1 (2013-12-11)

- Added menu-separator.
- Fixed the naming for the menu-item_hover.

## v0.18.0 (2013-12-11)

- Added menu skin.
- Added check icon (and reused it in the checkbox).

## v0.17.0 (2013-11-21)

- Added promo-button, #43.
- Added basic domik.

## v0.16.9 (2013-11-18)

- Fixed background for inputs in IE.

## v0.16.8 (2013-11-14)

- Fixed more IE valign problems.

## v0.16.7 (2013-11-14)

- In IE fallback to browser radios too.

## v0.16.6 (2013-11-14)

- Workaround for CSSO/IE error.
- In IE fallback to browser checkboxes.

## v0.16.5 (2013-11-13)

- Fixed rem on font problem in IE.

## v0.16.4 (2013-11-12)

- Fixed baseline problems with button in IE.

## v0.16.3 (2013-11-11)

- Fixed inputs' borders in IE, #42.

## v0.16.2 (2013-11-09)

- Fixed hardcoded disabled state for inputs.
- Fixed dropdown icon's padding, #41.
- Stylobate up.

## v0.16.1 (2013-11-06)

- Stylobate and Stylus deps up.

## v0.16.0 (2013-10-31)

- Updated Stylobate to the latest version.
- Moved variables to hashes.
- Added header text sizes.
- Fixed appearance of inputs, #38.

## v0.15.1 (2013-10-28)

- More shortcuts for variables in a standalone version.

- Shortcuts for variables are now lazy, so they won't break anything on include.

## v0.15.0 (2013-10-28)

- Better project structure, #37.
- `standalone.styl`, #37.
- Better keyword for atomic tests: `make test=button`.

## v0.14.4 (2013-10-22)

- Fixed search input appearance.

## v0.14.2-v0.14.3 (2013-09-25)

- Fixed namespace problems in popup and select.
- Updated Stylobate dep.

## v0.14.1 (2013-09-19)

- Updated Stylobate dep.

## v0.14.0

- Added new color and shadow variables.
- Refactored focus styles.
- Added input error style.

## v0.13.0 — v0.13.1

- Updated Stylobate dep.

## v0.12.0

- Now docstests are using CSSO and csscomb.js for pretier code.
- Changelog added.

## v0.11.0 — v0.11.4

- A lot of doctests added, some in-place refactoring.
- Added variables for simple backgrounds.
- Made misc color a bit brighter.
- Make the tail selectors less specific.

## v0.10.0 — v0.10.4

- Started to add proper doctests and Travis CI.
- Added selecty modifier to `icon dropdown`.
- Some refactoring and fixes for a lot of things.

## v0.9.0

- Added tail element for popup.

## v0.8.0 — v0.8.1

- Refactored popup.


rebuild-all-tests.js
/**
 * Module dependencies.
 */

var stylus = require('stylus');
var fs = require('fs');
var postcss = require('postcss');
var autoprefixer = require('autoprefixer');
var glob = require('glob');
var whatToTest = process.env.npm_package_config_whatToTest || '**';

// test cases

glob.sync("./lib/*/" + whatToTest + "/tests/*.styl").forEach(function(test){
  var name = test.replace(/\.?[\/]/g, ' ').replace(' tests',':').replace('.styl','');

  if(name.indexOf(' old_') != -1) {
    return;
  }

  var style = stylus('@require "node_modules/stylobate"; @require "lib/index"; @require "' + test + '"');
  style.render(function(err, css){
    if (err) throw err;

    var result = postcss()
        .use( autoprefixer({browsers: ['last 2 versions', 'ios 5', 'ie 9', 'fx 28', 'fx 21']}).postcss )
        .process(css);

    result.root.eachRule(function (rule) {
        // Insert an extra newline before each non-first rule
        if (rule.parent.rules[0] !== rule) {
            rule.before = '\n\n';
        }
    });

    css = result.css;
    // var css = postcss.parse(fs.readFileSync('test.css'));


    fs.writeFileSync(test.replace('.styl', '.css'), css);
  });
});



prism.js
/**
 * Prism: Lightweight, robust, elegant syntax highlighting
 * MIT license http://www.opensource.org/licenses/mit-license.php/
 * @author Lea Verou http://lea.verou.me
 */(function(){var e=/\blang(?:uage)?-(?!\*)(\w+)\b/i,t=self.Prism={util:{type:function(e){return Object.prototype.toString.call(e).match(/\[object (\w+)\]/)[1]},clone:function(e){var n=t.util.type(e);switch(n){case"Object":var r={};for(var i in e)e.hasOwnProperty(i)&&(r[i]=t.util.clone(e[i]));return r;case"Array":return e.slice()}return e}},languages:{extend:function(e,n){var r=t.util.clone(t.languages[e]);for(var i in n)r[i]=n[i];return r},insertBefore:function(e,n,r,i){i=i||t.languages;var s=i[e],o={};for(var u in s)if(s.hasOwnProperty(u)){if(u==n)for(var a in r)r.hasOwnProperty(a)&&(o[a]=r[a]);o[u]=s[u]}return i[e]=o},DFS:function(e,n){for(var r in e){n.call(e,r,e[r]);t.util.type(e)==="Object"&&t.languages.DFS(e[r],n)}}},highlightAll:function(e,n){var r=document.querySelectorAll('code[class*="language-"], [class*="language-"] code, code[class*="lang-"], [class*="lang-"] code');for(var i=0,s;s=r[i++];)t.highlightElement(s,e===!0,n)},highlightElement:function(r,i,s){var o,u,a=r;while(a&&!e.test(a.className))a=a.parentNode;if(a){o=(a.className.match(e)||[,""])[1];u=t.languages[o]}if(!u)return;r.className=r.className.replace(e,"").replace(/\s+/g," ")+" language-"+o;a=r.parentNode;/pre/i.test(a.nodeName)&&(a.className=a.className.replace(e,"").replace(/\s+/g," ")+" language-"+o);var f=r.textContent;if(!f)return;f=f.replace(/&/g,"&amp;").replace(/</g,"&lt;").replace(/\u00a0/g," ");var l={element:r,language:o,grammar:u,code:f};t.hooks.run("before-highlight",l);if(i&&self.Worker){var c=new Worker(t.filename);c.onmessage=function(e){l.highlightedCode=n.stringify(JSON.parse(e.data),o);t.hooks.run("before-insert",l);l.element.innerHTML=l.highlightedCode;s&&s.call(l.element);t.hooks.run("after-highlight",l)};c.postMessage(JSON.stringify({language:l.language,code:l.code}))}else{l.highlightedCode=t.highlight(l.code,l.grammar,l.language);t.hooks.run("before-insert",l);l.element.innerHTML=l.highlightedCode;s&&s.call(r);t.hooks.run("after-highlight",l)}},highlight:function(e,r,i){return n.stringify(t.tokenize(e,r),i)},tokenize:function(e,n,r){var i=t.Token,s=[e],o=n.rest;if(o){for(var u in o)n[u]=o[u];delete n.rest}e:for(var u in n){if(!n.hasOwnProperty(u)||!n[u])continue;var a=n[u],f=a.inside,l=!!a.lookbehind,c=0;a=a.pattern||a;for(var h=0;h<s.length;h++){var p=s[h];if(s.length>e.length)break e;if(p instanceof i)continue;a.lastIndex=0;var d=a.exec(p);if(d){l&&(c=d[1].length);var v=d.index-1+c,d=d[0].slice(c),m=d.length,g=v+m,y=p.slice(0,v+1),b=p.slice(g+1),w=[h,1];y&&w.push(y);var E=new i(u,f?t.tokenize(d,f):d);w.push(E);b&&w.push(b);Array.prototype.splice.apply(s,w)}}}return s},hooks:{all:{},add:function(e,n){var r=t.hooks.all;r[e]=r[e]||[];r[e].push(n)},run:function(e,n){var r=t.hooks.all[e];if(!r||!r.length)return;for(var i=0,s;s=r[i++];)s(n)}}},n=t.Token=function(e,t){this.type=e;this.content=t};n.stringify=function(e,r,i){if(typeof e=="string")return e;if(Object.prototype.toString.call(e)=="[object Array]")return e.map(function(t){return n.stringify(t,r,e)}).join("");var s={type:e.type,content:n.stringify(e.content,r,i),tag:"span",classes:["token",e.type],attributes:{},language:r,parent:i};s.type=="comment"&&(s.attributes.spellcheck="true");t.hooks.run("wrap",s);var o="";for(var u in s.attributes)o+=u+'="'+(s.attributes[u]||"")+'"';return"<"+s.tag+' class="'+s.classes.join(" ")+'" '+o+">"+s.content+"</"+s.tag+">"};if(!self.document){self.addEventListener("message",function(e){var n=JSON.parse(e.data),r=n.language,i=n.code;self.postMessage(JSON.stringify(t.tokenize(i,t.languages[r])));self.close()},!1);return}var r=document.getElementsByTagName("script");r=r[r.length-1];if(r){t.filename=r.src;document.addEventListener&&!r.hasAttribute("data-manual")&&document.addEventListener("DOMContentLoaded",t.highlightAll)}})();;
Prism.languages.markup={comment:/&lt;!--[\w\W]*?-->/g,prolog:/&lt;\?.+?\?>/,doctype:/&lt;!DOCTYPE.+?>/,cdata:/&lt;!\[CDATA\[[\w\W]*?]]>/i,tag:{pattern:/&lt;\/?[\w:-]+\s*(?:\s+[\w:-]+(?:=(?:("|')(\\?[\w\W])*?\1|\w+))?\s*)*\/?>/gi,inside:{tag:{pattern:/^&lt;\/?[\w:-]+/i,inside:{punctuation:/^&lt;\/?/,namespace:/^[\w-]+?:/}},"attr-value":{pattern:/=(?:('|")[\w\W]*?(\1)|[^\s>]+)/gi,inside:{punctuation:/=|>|"/g}},punctuation:/\/?>/g,"attr-name":{pattern:/[\w:-]+/g,inside:{namespace:/^[\w-]+?:/}}}},entity:/&amp;#?[\da-z]{1,8};/gi};Prism.hooks.add("wrap",function(e){e.type==="entity"&&(e.attributes.title=e.content.replace(/&amp;/,"&"))});;
Prism.languages.css={comment:/\/\*[\w\W]*?\*\//g,atrule:{pattern:/@[\w-]+?.*?(;|(?=\s*{))/gi,inside:{punctuation:/[;:]/g}},url:/url\((["']?).*?\1\)/gi,selector:/[^\{\}\s][^\{\};]*(?=\s*\{)/g,property:/(\b|\B)[\w-]+(?=\s*:)/ig,string:/("|')(\\?.)*?\1/g,important:/\B!important\b/gi,ignore:/&(lt|gt|amp);/gi,punctuation:/[\{\};:]/g};Prism.languages.markup&&Prism.languages.insertBefore("markup","tag",{style:{pattern:/(&lt;|<)style[\w\W]*?(>|&gt;)[\w\W]*?(&lt;|<)\/style(>|&gt;)/ig,inside:{tag:{pattern:/(&lt;|<)style[\w\W]*?(>|&gt;)|(&lt;|<)\/style(>|&gt;)/ig,inside:Prism.languages.markup.tag.inside},rest:Prism.languages.css}}});;
Prism.languages.css.selector={pattern:/[^\{\}\s][^\{\}]*(?=\s*\{)/g,inside:{"pseudo-element":/:(?:after|before|first-letter|first-line|selection)|::[-\w]+/g,"pseudo-class":/:[-\w]+(?:\(.*\))?/g,"class":/\.[-:\.\w]+/g,id:/#[-:\.\w]+/g}};Prism.languages.insertBefore("css","ignore",{hexcode:/#[\da-f]{3,6}/gi,entity:/\\[\da-f]{1,8}/gi,number:/[\d%\.]+/g,"function":/(attr|calc|cross-fade|cycle|element|hsla?|image|lang|linear-gradient|matrix3d|matrix|perspective|radial-gradient|repeating-linear-gradient|repeating-radial-gradient|rgba?|rotatex|rotatey|rotatez|rotate3d|rotate|scalex|scaley|scalez|scale3d|scale|skewx|skewy|skew|steps|translatex|translatey|translatez|translate3d|translate|url|var)/ig});;
Prism.languages.scss=Prism.languages.extend("css",{comment:{pattern:/(^|[^\\])(\/\*[\w\W]*?\*\/|\/\/.*?(\r?\n|$))/g,lookbehind:!0},atrule:/@[\w-]+(?=\s+(\(|\{|;))/gi,url:/([-a-z]+-)*url(?=\()/gi,selector:/([^@;\{\}\(\)]?([^@;\{\}\(\)]|&amp;|\#\{\$[-_\w]+\})+)(?=\s*\{(\}|\s|[^\}]+(:|\{)[^\}]+))/gm});Prism.languages.insertBefore("scss","atrule",{keyword:/@(if|else if|else|for|each|while|import|extend|debug|warn|mixin|include|function|return)|(?=@for\s+\$[-_\w]+\s)+from/i});Prism.languages.insertBefore("scss","property",{variable:/((\$[-_\w]+)|(#\{\$[-_\w]+\}))/i});Prism.languages.insertBefore("scss","ignore",{placeholder:/%[-_\w]+/i,statement:/\B!(default|optional)\b/gi,"boolean":/\b(true|false)\b/g,"null":/\b(null)\b/g,operator:/\s+([-+]{1,2}|={1,2}|!=|\|?\||\?|\*|\/|\%)\s+/g});
;
(function(){if(!self.Prism)return;var e=/\b([a-z]{3,7}:\/\/|tel:)[\w-+%~/.]+/,t=/\b\S+@[\w.]+[a-z]{2}/,n=/\[([^\]]+)]\(([^)]+)\)/,r=["comment","url","attr-value","string"];for(var i in Prism.languages){var s=Prism.languages[i];Prism.languages.DFS(s,function(i,s){if(r.indexOf(i)>-1){s.pattern||(s=this[i]={pattern:s});s.inside=s.inside||{};i=="comment"&&(s.inside["md-link"]=n);s.inside["url-link"]=e;s.inside["email-link"]=t}});s["url-link"]=e;s["email-link"]=t}Prism.hooks.add("wrap",function(e){if(/-link$/.test(e.type)){e.tag="a";var t=e.content;if(e.type=="email-link")t="mailto:"+t;else if(e.type=="md-link"){var r=e.content.match(n);t=r[2];e.content=r[1]}e.attributes.href=t}})})();;
(function(){if(!self.Prism||!self.document||!document.querySelector)return;var e={js:"javascript",html:"markup",svg:"markup"};Array.prototype.slice.call(document.querySelectorAll("pre[data-src]")).forEach(function(t){var n=t.getAttribute("data-src"),r=(n.match(/\.(\w+)$/)||[,""])[1],i=e[r]||r,s=document.createElement("code");s.className="language-"+i;t.textContent="";s.textContent="Loading…";t.appendChild(s);var o=new XMLHttpRequest;o.open("GET",n,!0);o.onreadystatechange=function(){if(o.readyState==4)if(o.status<400&&o.responseText){s.textContent=o.responseText;Prism.highlightElement(s)}else o.status>=400?s.textContent="✖ Error "+o.status+" while fetching file: "+o.statusText:s.textContent="✖ Error: File does not exist or is empty"};o.send(null)})})();;


script.js
// Shortcuts for Prism
Prism.languages.styl = Prism.languages.scss;
Prism.languages.html = Prism.languages.markup;

// Injecting source of examples' HTML
$('blockquote:not(:has(.example-code>.group)):not(:has(.example-source))').append($('<div class="example-code"><span class="group"></span></div>'))

$('blockquote:has(.example-code>.group)').each(function(){
    var $example_html = $(this).clone();
    $example_html.find('.example-code').remove();
    $child = $example_html.children();
    if ($child.length == 1 && $child[0].className == '') {
        $example_html = $child;
    }
    var example_html = $example_html[0].innerHTML.replace(/^\s+/,'').replace(/\s+$/,'');
    if (example_html) {
        var example_html_inner = example_html.split('\n');
        if (example_html_inner[example_html_inner.length - 1].match(/^\s\s\s\s/)) {
            example_html = example_html.replace(/\n    /g,'\n');
        }
        var $example_src = $(this).children('.example-code');
        var $example_src_code = $('<pre class="language-html is-hidden"><code></code></pre>');
        $example_src_code.children('code').text(example_html);
        $example_src.children('.group').append($('<span class="small-pseudo-button toggle-button group-item example-source js-toggler" data-toggle="html"><span class="button-content">html</span></span>'));
        $example_src.append($example_src_code);
    }
});


// Something for modals
$("[data-modal]").click(function(){
    $('#' + $(this).attr('data-modal')).toggleClass('is-hidden');
});
$(".modal-overlay, .modal-close").click(function(){
    $(this).closest('.modal-wrap').toggleClass('is-hidden');
});
$(".toggle-button:not(.is-disabled)").click(function(){
    $(this).toggleClass('is-checked');
});

// Toggling code blocks
$('.js-toggler:not([data-toggle])').click(function(){
    $(this).next().toggleClass('is-hidden');
});
$('.js-toggler[data-toggle]').click(function(){
    $(this).closest('.group').siblings('[class*="language-' + $(this).data('toggle') + '"]').toggleClass('is-hidden');
});
$('.js-outer-toggler').click(function(){
    $(this).parent().next().toggleClass('is-hidden');
});

// Toggling tabs
$('blockquote .tabs-item').click(function(){
    $(this).siblings().removeClass('tabs-item_selected');
    $(this).addClass('tabs-item_selected');
});

// Auto-sizing textarea and input

var reset_height = function(el) {
    el.height(0).height(el[0].scrollHeight);
}
var reset_width = function(el) {
    el.width(0).width(el[0].scrollWidth);
}
$('input.js-autosize').on('input',function(){
    reset_width($(this));
});
$('input.js-autosize').each(function(){
    reset_width($(this));
    $(this).closest('.field').css('width','auto');
});
$('textarea.js-autosize').on('input',function(){
    reset_height($(this));
});
$('textarea.js-autosize').each(function(){
    reset_height($(this));
});

$('.field-reset').on('click',function(e){
    e.preventDefault();
    var $field = $(this).closest('.field');
    var $input = $field.find('.input-controller');
    $input
        .val('')
        .trigger('input')
        .focus();
    if ($field.hasClass('js-autosize')) {
        $input
            .height(0)
            .height($input[0].scrollHeight);
    }
});

// Handling placeholder
$('.js-with-placeholder .input-controller').on('input', function(){
    var $input = $(this);
    var $field = $input.closest('.field, .small-field');
    var $placeholder = $field.find('.field-placeholder, .small-field-placeholder');

    if ($input.val() == '') {
        $placeholder.css('visibility', 'inherit');
    } else {
        $placeholder.css('visibility', 'hidden');
    }
}).trigger('input');

// Handling placeholder ghost
$('.js-with-placeholder-ghost .input-controller').on('input', function(){
    var $input = $(this);
    var $field = $input.closest('.field');
    var $ghost = $field.find('.field-placeholder-ghost');

    $ghost.text($input.val());
}).trigger('input');

$('.js-with-placeholder-ghost .input-controller').on('focus', function(){
    $(this)
        .closest('.field')
        .find('.field-placeholder')
        .css('visibility', 'hidden');
});

$('.js-with-placeholder-ghost .input-controller').on('blur', function(){
    var $input = $(this);
    if ($input.val() != '') {
        $input
            .closest('.field')
            .find('.field-placeholder')
            .css('visibility', 'inherit');
    }
});

// Handling placeholder autocomplete
$('.js-with-placeholder-autocomplete .input-controller').on('input', function(){
    var input = $(this);
    var field = input.closest('.field');
    var ghost = field.find('.field-placeholder-ghost');
    var autocomplete = field.find('.field-placeholder-content');
    var input_value = input.val();
    ghost.text(input_value);
    if ('Абракадабра'.indexOf(input_value) == 0) {
        autocomplete.text('Абракадабра'.replace(input_value, ''));
    } else {
        autocomplete.text('');
    }
}).trigger('input');

// Keeping the position of the sidebar

var lastScroll = $(window).scrollTop();

var topOffset = 20;

var elel = $('.aside');
var elel_is_fixed = false;
var elelHeight = elel.height();
var elelInitialTop = elel.offset().top;
var winHeight = $(window).height();
var maxScroll = $('body').height() - winHeight;

var bottomOffset = winHeight - elelHeight - 20;
var lastTop = Math.min(bottomOffset, 20);

$(window).resize(function() {
    winHeight = $(window).height();
    maxScroll = $('body').height() - winHeight;
    bottomOffset = winHeight - elelHeight - 20;
});

var checkScroll = function() {
    var winScroll = Math.min(Math.max($(window).scrollTop(), 0), maxScroll);

    if (winScroll > (elelInitialTop - lastTop - topOffset + 20)) {
        if (!elel_is_fixed) {
            elel_is_fixed = !elel_is_fixed;
            elel.addClass('aside_fixed');
            if (winHeight < elelHeight) {
                elel.css('top', bottomOffset);
                lastTop = bottomOffset;
            }
        }
    } else {
        if (elel_is_fixed) {
            elel_is_fixed = !elel_is_fixed;
            elel.removeClass('aside_fixed');
            if (winHeight < elelHeight) {
                elel.css('top', 'auto');
                lastTop = bottomOffset;
            }
        }
    }

    if (winHeight < elelHeight) {
        if (elel_is_fixed && winScroll > lastScroll) {
            if (lastTop > bottomOffset) {
                lastTop = Math.min(Math.max(lastTop + (lastScroll - winScroll), bottomOffset), topOffset);
                elel.css('top', lastTop);
            }
        } else if (elel_is_fixed && winScroll < lastScroll) {
            if (lastTop < topOffset) {
                lastTop = Math.min(Math.max(lastTop + (lastScroll - winScroll), bottomOffset), topOffset);
                elel.css('top', lastTop);
            }
        }
    }
    lastScroll = winScroll;

}

checkScroll();
$(window).scroll(checkScroll);


test-runner.js
/**
 * Module dependencies.
 */

var stylus = require('stylus');
var fs = require('fs');
var postcss = require('postcss');
var autoprefixer = require('autoprefixer');
var glob = require('glob');
var whatToTest = process.env.npm_package_config_whatToTest || '**';

// test cases

glob.sync("./lib/*/" + whatToTest + "/tests/*.styl").forEach(function(test){
  var name = test.replace(/\.?[\/]/g, ' ').replace(' tests',':').replace('.styl','');
  if(name.indexOf(' old_') != -1) {
    return;
  }

  it(name, function(){
    var css = fs.readFileSync(test.replace('.styl', '.css'), 'utf8').replace(/\r/g, '').trim();
    var style = stylus('@require "node_modules/stylobate"; @require "lib/index"; @require "' + test + '"');

    style.render(function(err, actual){
      if (err) throw err;

      actual = postcss()
          .use( autoprefixer({browsers: ['last 2 versions', 'ios 5', 'ie 9', 'fx 28', 'fx 21']}).postcss )
          .process(actual);

      actual.root.eachRule(function (rule) {
          // Insert an extra newline before each non-first rule
          if (rule.parent.rules[0] !== rule) {
              rule.before = '\n\n';
          }
      });

      actual.css.trim().should.equal(css);
    });
  })
});

